tests/acceptance/test_sasl.py

```diff
@@ -0,0 +1,71 @@
+# MIT License
+# Copyright (c) 2020-2024 Pau Freixes
+
+import asyncio
+import logging
+import os
+import sys
+from pathlib import Path
+
+import pytest
+
+from emcache import MemcachedHostAddress, create_client
+
+pytestmark = pytest.mark.asyncio
+
+logging.basicConfig(level=logging.DEBUG)
+
+with open(Path(os.path.dirname(__file__)) / "data" / "auth_pwd.txt", "r") as f:
+    username, password = f.readline().rstrip().split(":")
+
+
+@pytest.fixture()
+async def authed_client(event_loop):
+    client = await create_client(
+        [MemcachedHostAddress("localhost", 11214)], timeout=1.0, username=username, password=password
+    )
+    try:
+        yield client
+    finally:
+        await client.close()
+
+
+class TestAuth:
+    @pytest.mark.skipif(sys.platform == "darwin", reason="This server is not built with SASL support.")
+    async def test_auth(self, authed_client):
+        await authed_client.get(b"key")
+
+    @pytest.mark.skipif(sys.platform == "darwin", reason="This server is not built with SASL support.")
+    async def test_auth_with_errors(self):
+        client_no_userpass = await create_client([MemcachedHostAddress("localhost", 11214)], timeout=1.0)
+        with pytest.raises(asyncio.TimeoutError):
```

Comment:
Thanks for adding this test.

This is how our driver will behave in case of having the wrong credentials configured, or not using them at all. The only hint that the user will have that something is off will be the warning logs emitted by the background job for creating a new connection.

Wondering if we could bubble up this error more explicitly in the sense that we could have a test like this:

```
    async def test_auth_with_errors(self):
        client_no_userpass = await create_client([MemcachedHostAddress("localhost", 11214)], timeout=1.0)
        with pytest.raises(emcache.AuthenticationError):
```

The good part of this approach for the interface would be that, first you dont need to relay on the `Timeout` for having the client noticing that the operation will never be made, and second that tihs makes the error explicit.

For achieving this, wondering if we could change a bit this [one](https://github.com/emcache/emcache/blob/master/emcache/connection_pool.py#L215), so in case of an `auth` error we setup the exception for the waiter `set_exception` with the auth exception, then [here](https://github.com/emcache/emcache/blob/master/emcache/connection_pool.py#L439) we just need to catch this exception and bubble up this to the user.

For this working we will need to remove also the usage of the `backoff` strategy due to the `error` configured to `True`.

WDYT?

Reply:
Thank you for explicitly pointing out exactly where this implicit expectation can be fixed. Since this is exactly what I was trying to achieve when I initially wanted to throw out an obvious error, but did not understand at what level it was necessary to fix it. This timeout has completely hidden everything.

Reply:
I've looked at this option and I don't understand how you want to get an error when creating a client if the changes you propose only work when accessing memcached commands. in order for the error to occur when creating a client, changes are needed at the create_client level. You've got me confused.

Reply:
did you mean that the error was due to a lack of parameters or an authentication error?

Reply:
We will have two type of errors, the ones that are returned during the instantiation of the client which will be the ones related to provide mostly invalid parameters, for example setting `username` but not `password` or the other way arround.

If the user set both of them then we will have errors at memcached command level if for example the the credentials were wrong. By doing this we can explicitly bubble up to the client an error due to authentication while we do not relay on having to expect from the client to configure a timeout, otherwise the user if would not configure a timeout for the client the operation will never finish. However by bubbling up - using the `set_exception`- we can quickly respond to the user that something wrong happened with the command, and explicitly raising an Auth exception.

Let me know if Im not explaining properly or you still have some oustanding questions.

Reply:
> did you mean that the error was due to a lack of parameters or an authentication error?
As said the lack of parameters, because only one was set will be raised at client instantiation. Other errors associated to bad credentials usage will be raise during the usage of memcache commands

Reply:
I saw that you need to remove the error at the connection pool level
```bash
DEBUG:asyncio:Using selector: EpollSelector
DEBUG:emcache.node:<Node host=127.0.0.1 port=11214 closed=False> new node created
INFO:emcache.cluster:Nodes used for sending traffic: [<Node host=127.0.0.1 port=11214 closed=False>]
DEBUG:emcache.cluster:Cluster configured with 1 nodes
DEBUG:emcache.cluster:Events dispatcher started
DEBUG:emcache.protocol:Connection made
INFO:emcache.connection_pool:<ConnectionPool host=127.0.0.1 port=11214 total_connections=1 min_connections=1 max_connections=2 closed=False> new connection created

```

Reply:
Stay tuned for updated PR!)))

Reply:
There's just one small question left. As I understand you, you want to remove access to the server through the version command. However, if the username and password are empty, we will not attempt to authenticate and will receive a timeouterror. To get an autherror when the username and password arguments are empty, we must try either through the version command to find out what status the server has (whether it supports SASL) or with each new connection, if the username and password arguments are missing, they will try to authenticate. Otherwise, if we reject all this in the absence of arguments for authentication with the client.get command, we will receive a timeouterror and will not even be able to set the autherror error since we have not even tried. How do you think?

Reply:
Let's just send what I have. I apologize for the frequent clarifications. The feature is not the easiest for me. thanks for your help. Every detail is very important to me. So I get a better and better idea of ​​the feature. Thank you)

Reply:
Can we create different tests for each use case? will help on making them more clear.

Regarding this one:

> Otherwise, if we reject all this in the absence of arguments for authentication with the client.get command, we will receive a timeouterror and will not even be able to set the autherror error since we have not even tried. How do you think?

If I understood well there is another use case where auhentication is required but the user did not configure any user and password, is that correct?

Then my question here, is what the memcache server will do if we try to make a `.get` (or any other command) and the user did not first get authenticated? will just ignore the command or will return an error? if its the later I would say that we are fine. right?

Reply:
Well, of course, you just need to return the error that memcached returns! Thank you

Reply:
Resolving this for allowing me to start a new thread of conversations in a cleaner state